P027 乳腺癌数据集 - 加载并认识数据

In [1]:
from sklearn.datasets import load_breast_cancer

打印数据集的详细描述信息，以便于了解数据集的背景、特征和使用情况
数据集的背景和用途。
数据的来源（如是由某些研究机构提供）。
样本数量、特征数量及每个特征的含义。
数据集中的每个类别及其对应的标签。

In [3]:
breast_cancer = load_breast_cancer()
print(breast_cancer["DESCR"])

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

P028 查看data和target

In [4]:
data = breast_cancer["data"]
target = breast_cancer["target"]

In [5]:
data.shape

(569, 30)

In [6]:
target.shape

(569,)

In [7]:
type(data),type(target)

(numpy.ndarray, numpy.ndarray)

In [8]:
target[:3]

array([0, 0, 0])

P029 合并data和target到大数组

In [9]:
import numpy as np

In [10]:
data.shape

(569, 30)

In [11]:
target.shape

(569,)

np.c_[] 是 NumPy 中的一种简洁的拼接方式，用于将多个数组沿列方向（即水平方向）拼接在一起。
在这个例子中，data 是一个形状为 (569, 30) 的二维数组，包含乳腺癌数据集的特征。
target 是一个形状为 (569,) 的一维数组，包含每个样本的分类标签（目标值）。
通过 np.c_[data, target]，将 target 作为一个新的列拼接到 data 的右边，形成一个新的二维数组 all_datas。

In [12]:
all_datas = np.c_[data,target]

In [13]:
all_datas.shape

(569, 31)

In [14]:
all_datas[:3]

array([[1.799e+01, 1.038e+01, 1.228e+02, 1.001e+03, 1.184e-01, 2.776e-01,
        3.001e-01, 1.471e-01, 2.419e-01, 7.871e-02, 1.095e+00, 9.053e-01,
        8.589e+00, 1.534e+02, 6.399e-03, 4.904e-02, 5.373e-02, 1.587e-02,
        3.003e-02, 6.193e-03, 2.538e+01, 1.733e+01, 1.846e+02, 2.019e+03,
        1.622e-01, 6.656e-01, 7.119e-01, 2.654e-01, 4.601e-01, 1.189e-01,
        0.000e+00],
       [2.057e+01, 1.777e+01, 1.329e+02, 1.326e+03, 8.474e-02, 7.864e-02,
        8.690e-02, 7.017e-02, 1.812e-01, 5.667e-02, 5.435e-01, 7.339e-01,
        3.398e+00, 7.408e+01, 5.225e-03, 1.308e-02, 1.860e-02, 1.340e-02,
        1.389e-02, 3.532e-03, 2.499e+01, 2.341e+01, 1.588e+02, 1.956e+03,
        1.238e-01, 1.866e-01, 2.416e-01, 1.860e-01, 2.750e-01, 8.902e-02,
        0.000e+00],
       [1.969e+01, 2.125e+01, 1.300e+02, 1.203e+03, 1.096e-01, 1.599e-01,
        1.974e-01, 1.279e-01, 2.069e-01, 5.999e-02, 7.456e-01, 7.869e-01,
        4.585e+00, 9.403e+01, 6.150e-03, 4.006e-02, 3.832e-02, 2.058e-02

P030 乳腺癌数据集 - 生成Pandas的df

In [15]:
breast_cancer["feature_names"]

array(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error',
       'fractal dimension error', 'worst radius', 'worst texture',
       'worst perimeter', 'worst area', 'worst smoothness',
       'worst compactness', 'worst concavity', 'worst concave points',
       'worst symmetry', 'worst fractal dimension'], dtype='<U23')

In [17]:
import pandas as pd

np.c_[data, target] 将特征数据 data 和目标值 target 拼接在一起，形成一个新的二维数组。每一行对应一个样本，前面是特征值，最后一列是目标值。

columns=list(breast_cancer["feature_names"]) + ["target"] 指定了数据框的列名。首先将 breast_cancer 数据集中的特征名（feature_names）转换为一个列表，然后加上 ["target"]，表示最后一列是目标值。

In [18]:
df = pd.DataFrame(
    np.c_[data, target],
    columns = list(breast_cancer["feature_names"]) + ["target"]
)

In [19]:
df.head(3)

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38           122.8     1001.0          0.11840   
1        20.57         17.77           132.9     1326.0          0.08474   
2        19.69         21.25           130.0     1203.0          0.10960   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33            184.6      2019.0   
1                 0.05667  ...          23.41            158.8      1956.0   
2                 0.05999  ...          25.53            152.5      1709.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   

   worst symmetry  worst fractal dimension  target  
0          0.4601                  0.11890     0.0  
1          0.2750                  0.08902     0.0  
2          0.3613                  0.08758     0.0  

[3 rows x 31 columns]

P031 乳腺癌数据集 - 拆分训练集和测试集

In [20]:
from sklearn.model_selection import train_test_split

data：输入数据，即特征矩阵。
target：目标标签，即分类标签。
random_state=40：随机种子，保证每次拆分的结果一致。
test_size=0.25：测试集占整个数据集的比例，即25%。
将数据集拆分为训练集和测试集，并指定测试集的大小和随机种子，以保证拆分结果的可重复性

In [21]:
X_train, X_test, y_train, y_test = train_test_split(
    data, target, random_state=40, test_size=0.25)

In [22]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((426, 30), (143, 30), (426,), (143,))

P032 乳腺癌数据集 - 训练测试集数据分布

In [23]:
target.shape, y_train.shape, y_test.shape

((569,), (426,), (143,))

通过 zip 函数将目标标签的名称和数据打包在一起，然后循环遍历每一个名称和数据对。
对于每一个数据对，打印数据的名称，并使用 pandas.Series 和 value_counts 方法计算并打印每个值在数据中的比例。

In [24]:
for name, array in zip(
    ["target", "y_train", "y_test"], [target, y_train, y_test]):
    print()
    print(name)
    print(pd.Series(array).value_counts(normalize=True))


target
1    0.627417
0    0.372583
Name: proportion, dtype: float64

y_train
1    0.607981
0    0.392019
Name: proportion, dtype: float64

y_test
1    0.685315
0    0.314685
Name: proportion, dtype: float64


P033 乳腺癌数据集 - 训练测试集的均匀拆分

In [25]:
X_train, X_test, y_train, y_test = train_test_split(
    data, target, random_state=40, test_size=0.25, stratify=target)

In [26]:
for name, array in zip(
    ["target", "y_train", "y_test"], [target, y_train, y_test]):
    print()
    print(name)
    print(pd.Series(array).value_counts(normalize=True))


target
1    0.627417
0    0.372583
Name: proportion, dtype: float64

y_train
1    0.626761
0    0.373239
Name: proportion, dtype: float64

y_test
1    0.629371
0    0.370629
Name: proportion, dtype: float64
